client/ting.jsx

```diff
@@ -38,57 +34,28 @@ const Ting = React.createClass({
 
         return {
             channel,
-            intendedUsername: null,
             // TODO(dionyziz): race conditions and queues
-            currentMessageId: null
         };
     },
     componentWillMount() {
-        const URL = window.location.hostname + ':' + config.port;
-        this._socket = io.connect(URL, {secure: config.websocket.secure});
-
-        this._socket.on('login-response', ({success, people, error}) => {
-            if (!success) {
-                this.refs.loginForm.onError(error);
-            }
-            else {
-                this.refs.loginForm.onSuccess();
-
-                var peopleList = _.chain(people)
-                    .without(this.intendedUsername)
-                    .value();
-
-                this.onLogin(this.state.intendedUsername, peopleList);
-            }
-        });
-
-        this._socket.on('message', (data) => {
-            this.refs.history.onMessage(data);
-        });
-
-        this._socket.on('part', (username) => {
-            this.refs.userList.onPart(username);
-            this.refs.history.deleteTypingMessage(username);
-        });
-
-        this._socket.on(
-            'join',
-            (username) => this.refs.userList.onJoin(username)
-        );
-
-        this._socket.on('start-typing-response', (messageid) => {
-            this.setState({currentMessageId: messageid});
-            this.refs.messageForm.onStartTypingResponse(messageid);
-        });
-
-        this._socket.on('update-typing-messages', (messagesTyping) => {
-            this.refs.history.onUpdateTypingMessages(messagesTyping);
-        });
-
         Analytics.init();
     },
+    componentDidMount() {
+        this.props.listenOnLoginResponse(this.refs.loginForm.onError, this.refs.loginForm.onSuccess, this.onLogin);
+        this.props.listenOnMessage(this.refs.history.onMessage);
+        this.props.listenOnPart(this.refs.userList.onPart, this.refs.history.deleteTypingMessage);
+        this.props.listenOnJoin(this.refs.userList.onJoin);
+        this.props.listenOnStartTypingResponse(this.refs.messageForm.onStartTypingResponse);
+        this.props.listenOnUpdateTypingMessages(this.refs.history.onUpdateTypingMessages);
```

Comment:
This is causing a memory leak, because you're not cleaning up the event registrations. You need to create `unlisten` functions on the parent Component which invoke socket.io's [removeListener](https://stackoverflow.com/questions/9418697/how-to-unsubscribe-from-a-socket-io-subscription). Then, you must call these from [componentWillUnmount](https://facebook.github.io/react/docs/react-component.html#componentwillunmount) in this component.

This should be similar to the way you unregister the keyDown event in the Setting's `componentDidUnmount` method.